src/mscorlib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -20,7 +20,34 @@ internal static partial class AsyncMethodBuilder
             {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.stateMachine);
             }
+#if CORERT // doesn't use try/finally, so no need for extra enregistrering work and no finally cloning
```

Comment:
There is not really a point in sharing the file when it is under one giant ifdef.

Reply:
The main practical difference between the two is coreclr uses try/finally and corert doesn't. 

I have 3 questions in this area:

1. Should the two differ in this respect?
2. If not, which one is the correct approach?
3. Is it a correct assumption that statemachines are unique per method; so force inlining `Start` wouldn't cause any duplicated code?

Reply:
> Should the two differ in this respect?

No

> If not, which one is the correct approach?

try/finally is correct.  The question is whether it's worth forgoing that correctness (which only matters in niche corner cases) in the name of performance, which is why I asked: 
https://github.com/dotnet/coreclr/pull/16124#issuecomment-361997706

> Is it a correct assumption that statemachines are unique per method

Technically, no, as you can create a state machine type that's a class rather than a struct, and the implementation will be shared.  In fact, that's what the C# compiler does for debug builds, as a class makes Edit-and-Continue possible in a way that's not possible or too challenging with a struct.  It's also possible for someone to hand write their own state machine object that's a class.  In practice, in release builds where you care about perf, the 99.9% case is they'd be structs, in which case in coreclr I'd expect them not to be shared.  But I believe in .NET Native / CoreRT there are also fallbacks in place for sharing code even for structs when there are too many instantiations; but Jan would know much better about that and whether it applies here.

Reply:
Ok, will do a performance analysis